src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatProvider.cs

```diff
@@ -0,0 +1,328 @@
+                    {
+                        setResult = false;
+                        throw new Exception("Cannot set a health property without adding it first.");
+                    },
+                    (key, property) =>
+                    {
+                        if (isHealthy != null)
+                        {
+                            property.IsHealthy = isHealthy.Value;
+                        }
+                        if (healthPropertyValue != null)
+                        {
+                            property.PayloadValue = healthPropertyValue;
+                        }
+
+                        return property;
+                    });
+
+                    return setResult;
+                }
+                catch (Exception e)
+                {
+                    CoreEventSource.Log.FailedToSetHeartbeatProperty(healthPropertyName, healthPropertyValue, isHealthy.HasValue ? isHealthy.Value.ToString() : "null", e.ToString());
+                }
+            }
+            else
+            {
+                CoreEventSource.Log.FailedToSetHeartbeatProperty(healthPropertyName, healthPropertyValue, isHealthy.HasValue ? isHealthy.Value.ToString() : "null");
+            }
+
+            return false;
+        }
+
+        #region IDisposable Support
+
+        public void Dispose()
+        {
+            this.Dispose(true);
+        }
+
+        /// <summary>
+        /// Get the metric telemetry item that will be sent.
+        /// 
+        /// Note: exposed to internal to allow inspection for testing.
+        /// </summary>
+        /// <returns>A MetricTelemtry item that contains the currently defined payload for a heartbeat 'pulse'.</returns>
+        internal ITelemetry GatherData()
+        {
+            var hbeat = new MetricTelemetry(heartbeatSyntheticMetricName, 0.0);
+
+            string updatedKeys = string.Empty;
+            string comma = string.Empty;
+
+            if (hbeat.Properties.ContainsKey(HeartbeatDefaultPayload.UpdatedFieldsPropertyKey))
+            {
+                updatedKeys = hbeat.Properties[HeartbeatDefaultPayload.UpdatedFieldsPropertyKey];
+                comma = ",";
+            }
+
+            foreach (var payloadItem in this.heartbeatProperties)
+            {
+                hbeat.Properties.Add(payloadItem.Key, payloadItem.Value.PayloadValue);
+                hbeat.Sum += payloadItem.Value.IsHealthy ? 0 : 1;
+                if (payloadItem.Value.IsUpdated)
+                {
+                    string.Concat(updatedKeys, comma, payloadItem.Key);
+                    comma = ",";
+                }
+            }
+
+            // update the special 'updated keys' property with the names of keys that have been updated.
+            if (!string.IsNullOrEmpty(updatedKeys))
+            {
+                hbeat.Properties[HeartbeatDefaultPayload.UpdatedFieldsPropertyKey] = updatedKeys;
```

Comment:
what is this functionality for? I do not remember it in previous PR and discussion of it. Seems like a redundant data

Reply:
Taking it out for this PR.
This is in my ‘Potential Fields’ in my design doc. I was inspired by the Geneva Heartbeat QoS incarnationId field, but wanted to make it more specific to unhealthy state changes only, and immediately useable in a UX scenario. Let’s discuss this as a separate issue + PR (I will bring this up with you after this PR is complete, I've put the changes into a separate branch in my own repo for now).